### PR TITLE
[docs][guide] approaches to writing integrations

### PR DESCRIPTION
**Pull request by cmpadden**
- Introduces a landing page for guides on writing integrations, currently outlined include:
    * asset factory
    * multi-asset decorator
    * resource definitions
    * pipes

### CODE REVIEW

**Comment by cmpadden** [comment]
* **#22903** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22903?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @cmpadden and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+### Cons
+
+- **Limited abstraction** While the resource can be re-used throughout the codebase, it does not provide any higher level abstraction to assets or jobs.
+
+### Tutorial
+
+<Note>A tutorial for writing a resource-based integration is coming soon!</Note>
+
+## Builder methods
```
Comment (PedramNavid) [inline docs/content/guides/integrations/approaches-to-writing-integrations.mdx:31]: Needs a paragraph that describes what a Builder method is. An example code block could be helpful

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+There are many approaches to writing integrations in Dagster. The choice of approach depends on the specific requirements of the integration, the level of control needed, and the complexity of the external system being integrated. By reviewing the pros and cons of each approach, it is possible to make an informed decision on the best method for a specific use case. The following are typical approaches that align with Dagster's best practices.
+
+- Resource providers
+- Builder methods
+- Multi-Asset decorators
+- Pipes protocol
+
+## Resource providers
+
+One of the most fundamental features that can be implemented in an integration is a resource object to interface with an external service. For example, the `dagster-snowflake` integration provides a custom `SnowflakeResource` that is a wrapper around the Snowflake `connector` object.
```
Comment (PedramNavid) [inline docs/content/guides/integrations/approaches-to-writing-integrations.mdx:16]: Could be helpful to link to the implementation of the resource.

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+- **Complexity:** Can be more complex to set up compared to other methods.
+- **Boilerplate code:** May require more boilerplate code to define assets, resources, and jobs.
+
+### Tutorial
+
+<Note>A tutorial for writing a builder method integration is coming soon!</Note>
+
+## Multi-asset decorators
+
```
Comment (PedramNavid) [inline docs/content/guides/integrations/approaches-to-writing-integrations.mdx:49]: Same comment as above re: paragraph

**Review by PedramNavid — DISMISSED** [review]

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+- **Simple** Implementing a resource wrapper is often the first step in flushing out a fully-featured integration.
+- **Reusable** Resources are a core building block in the Dagster ecosystem, and allow one to re-use code across assets.
+
+### Cons
+
+- **Limited abstraction** While the resource can be re-used throughout the codebase, it does not provide any higher level abstraction to assets or jobs.
+
+### Tutorial
+
+<Note>A tutorial for writing a resource-based integration is coming soon!</Note>
```
[the remark this answers is not in the document]
Comment (cmpadden) [reply docs/content/guides/integrations/approaches-to-writing-integrations.mdx:29]: Renamed all occurrences of tutorial -> guides

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx (hunk 1 above)
Comment (cmpadden) [reply docs/content/guides/integrations/approaches-to-writing-integrations.mdx:31]: Added intro paragraph along with example

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+One of the most fundamental features that can be implemented in an integration is a resource object to interface with an external service. For example, the `dagster-snowflake` integration provides a custom [SnowflakeResource](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-snowflake/dagster_snowflake/resources.py) that is a wrapper around the Snowflake `connector` object.
+
+### Pros
+
+- **Simple** Implementing a resource wrapper is often the first step in flushing out a fully-featured integration.
+- **Reusable** Resources are a core building block in the Dagster ecosystem, and allow one to re-use code across assets.
+
+### Cons
+
+- **Limited abstraction** While the resource can be re-used throughout the codebase, it does not provide any higher level abstraction to assets or jobs.
```
Comment (yuhan) [inline docs/content/guides/integrations/approaches-to-writing-integrations.mdx:25]:
```suggestion
- **Low-level abstraction** While the resource can be re-used throughout the codebase, it does not provide any higher level abstraction to assets or jobs.
```

"Limited" can mean either very low-level (lack of built-in support) or very high-level (lack of composability/flexibility). So suggest rewording this to make the limitation very explicit.

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+### Guide
+
+<Note>A guide for writing a resource based integration is coming soon!</Note>
+
+## Factory methods
```
Comment (yuhan) [inline docs/content/guides/integrations/approaches-to-writing-integrations.mdx:31]: Which integration do you have in mind? I don't think we have any builtin integrations are written this way (the good ones). If we're just talking about how a user can build their own integration in factory, I think that's fine but maybe we could move it down to a "Other patterns".

Stepping back, I think this guide will have two audience:
1. Partners / Contributors who want to build integrations that other users in the public would use
2. Individual organizations/users who want to integrate Dagster with their tech stack, specific to their needs.

To me, factory methods fit 2 but not 1 because it limits the composability. Therefore, I thought maybe we could move it to a separate section and keep the 1 with those other blessed patterns.

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+- **Complexity in debugging:** Debugging issues can be more challenging when multiple assets are defined in a single function.
+
+### Guide
+
+<Note>
+  A guide for writing a multi-asset decorator based integration is coming soon!
+</Note>
+
+## Pipes protocol
+
```
Comment (yuhan) [inline docs/content/guides/integrations/approaches-to-writing-integrations.mdx:102]: Pipes seems be missing the short explainer. I took a stab:
> The Pipes protocol is used to integrate with systems that have their own execution environments. It enables running code in these external environments while allowing Dagster to maintain control and visibility. Example implementations of this approach include [AWS Lambda](https://github.com/dagster-io/dagster/tree/master/python_modules/libraries/dagster-aws/dagster_aws/pipes), [Databricks](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-databricks/dagster_databricks/pipes.py), and [Kubernetes](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-k8s/dagster_k8s/pipes.py).

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+- **Streaming logs and metadata:** Provides support for streaming logs and structured metadata back into Dagster.
+
+### Cons
+
+- **Complexity:** Can be complex to set up and configure.
+- **Overhead:** May introduce additional overhead for managing external environments.
+
+### Guide
+
+<Note>A guide for writing a pipes based integration is coming soon!</Note>
```
Comment (yuhan) [inline docs/content/guides/integrations/approaches-to-writing-integrations.mdx:116]: can we link to this? https://docs.dagster.io/concepts/dagster-pipes/dagster-pipes-details-and-customization

or, do want to write a more hands-on tutorial like piece?

**Review by yuhan — COMMENTED** [review]
apologies for the long overdue review. this looks amazing. i left some suggestions in comments, but overall lg(great)2m

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx (hunk 8 above)
Comment (cmpadden) [reply docs/content/guides/integrations/approaches-to-writing-integrations.mdx:116]: I've included that link for now, thanks.

File: docs/content/guides/integrations/approaches-to-writing-integrations.mdx (hunk 6 above)
Comment (yuhan) [reply docs/content/guides/integrations/approaches-to-writing-integrations.mdx:31]: lets address this as a follow up!

**Review by yuhan — APPROVED** [review]

**Comment by cmpadden** [comment]
### Merge activity

* **Aug 23, 2:46 PM EDT**: @cmpadden started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/22903).
